### ISSUE TITLE
Please do not remove Public API without proper deprecation cycle

### ISSUE THREAD

**Issue by phako** [Exiv2/exiv2#611]
With https://github.com/Exiv2/exiv2/pull/601/commits/427bbad060f044f5857416af6633f3595af198ba, Exiv::Image::AutoPtr got replaced with Exiv::Image::UniquePtr. It is very hard to follow stable and "next" versions if you do that.

Also, from semantics, I don't think that there is even a need to do the rename on Exiv2 level. It's still an automatic pointer for Exiv2::Image.

So either, please make it throw a deprecation warning in 0.27 and remove it in 0.28 or re-add the type as a typedef to ::UniquePtr and mark it deprecated in 0.28, remving the typedef in 0.29

**Comment by clanmills** [comment]
Exiv2 v0.28 will deprecate several features on Exiv2.  The code will be "modernised" to C++11.

To respect the existing users of the current API, we plan to offer LTS (Long Term Support) for Exiv2 v0.27 with dot releases (for security fixes) in 2019 and 2020.

I'm not sure that I have understood your request for a "deprecation warning in 0.27".  @piponazo is  Lead Engineer for v0.28 and I will do the release engineering for the 0.27 "dots".

**Comment by D4N** [comment]
Adding a typedef that's marked as deprecated won't hurt, I think we can do that.

**Comment by cgilles** [comment]
MXE env take 2/3 hours maximum to compile from scratch here, but for digiKam with the huge list of packages. I use a simple ADSL connection here and a i7 cpu + 32 Gb RAM.

To compile only Exiv2 it take 30 mns, if you use the script that i shared few weeks ago.

Best

Gilles

**Comment by piponazo** [comment]
Hi guys. Thanks for bringing this topic to discussion. Although we have been discussing about migrating to c++11 for almost a year (see #214), it is true that we did not discuss about a proper deprecation cycle. 

I think I prefer the idea of adding a deprecation warning in the 0.27 branch to be able to develop in master (0.28) without too much restrictions. 

Regarding the name change from `Exiv::Image::AutoPtr` to `Exiv::Image::UniquePtr` I also agree that we could keep the previous name to not make so many changes in the API. If that helps I am happy to change back the name of the definition.

**Comment by phako** [comment]
Actually, due to the slight differences in std::unique_ptr and std::auto_ptr, It's probably more confusing to leave the name the same. I.e. for passing function parameters by value, you explicitly have to transfer the ownership with std::unique_ptr. 

Also, yes, with proper deprecation I meant using compiler attributes such as the ones you presented. Note that those only work for gcc and clang. There are more generic ones being part of C++14  but that is a bit moot if you're doing the porting porting to later versions (see https://en.cppreference.com/w/cpp/language/attributes)

**Comment by piponazo** [comment]
Note that in #617 I proposed a new mechanism to deprecate things on the 0.27 branch. This is using the proper way to deprecate things with CMake, that will generate a macro using compiler attributes or other mechanism depending on the compiler and standard used.

**Comment by piponazo** [comment]
@phako We have merged #617 into 0.27 so that we will have a new deprecation mechanism in place for the stable releases from now on. 

I still need to mark as deprecated `Exiv::Image::AutoPtr` which was the main reason why you came here in the first place. Please, let us know your thoughts about the system we just have added, and if you are happy about it, do not forget to close the ticket 😉

**Comment by phako** [comment]
We seem to have some issues with that MR:

```
ccache c++ -Igexiv2/1a6c516@@gexiv2@sha -Igexiv2 -I../../../src/gnome3/gexiv2/gexiv2 -Igexiv2/.. -I../../../src/gnome3/gexiv2/gexiv2/.. -I/home/valabot/gnome3/include/glib-2.0 -I/home/valabot/gnome3/lib/glib-2.0/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++11 -O2 -g -fPIC -pthread  -MD -MQ 'gexiv2/1a6c516@@gexiv2@sha/gexiv2-metadata.cpp.o' -MF 'gexiv2/1a6c516@@gexiv2@sha/gexiv2-metadata.cpp.o.d' -o 'gexiv2/1a6c516@@gexiv2@sha/gexiv2-metadata.cpp.o' -c ../../../src/gnome3/gexiv2/gexiv2/gexiv2-metadata.cpp
In file included from /home/valabot/gnome3/include/exiv2/exiv2.hpp:34,
                 from ../../../src/gnome3/gexiv2/gexiv2/gexiv2-metadata-private.h:14,
                 from ../../../src/gnome3/gexiv2/gexiv2/gexiv2-metadata.cpp:12:
/home/valabot/gnome3/include/exiv2/basicio.hpp:30:10: fatal error: exiv2lib_compiler_detection.h: No such file or directory
 #include "exiv2lib_compiler_detection.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by phako** [comment]
Looks fine. thanks. Btw. I would also merge the mechanism to master.
